### ISSUE TITLE
Postgres support

### ISSUE THREAD

**Issue by DangerOnTheRanger** [DangerOnTheRanger/maniwani#13]
This is just a formalization of work already in-progress since @9034725985 has been working on Postgres support (kudos for that!) so it can be properly tracked and closed with the `production-ready` milestone.

**Comment by 9034725985** [comment]
We will need tests to get production ready but here's what I see if I try to use postgresql today (openroom is the name of the database on postgresql on my local computer) 

declaring the database connection looks something like 
```
app.config["SQLALCHEMY_DATABASE_URI"] = "postgres://postgres@127.0.0.1:5432/openroom"
```

The format is postgres://username:password@hostname:port/databasename


```
openroom=# \d

              List of relations
 Schema |     Name      |   Type   |  Owner   
--------+---------------+----------+----------
 public | board         | table    | postgres
 public | board_id_seq  | sequence | postgres
 public | media         | table    | postgres
 public | media_id_seq  | sequence | postgres
 public | post          | table    | postgres
 public | post_id_seq   | sequence | postgres
 public | poster        | table    | postgres
 public | poster_id_seq | sequence | postgres
 public | reply         | table    | postgres
 public | session       | table    | postgres
 public | slip          | table    | postgres
 public | slip_id_seq   | sequence | postgres
 public | tag           | table    | postgres
 public | tag_id_seq    | sequence | postgres
 public | tags          | table    | postgres
 public | thread        | table    | postgres
 public | thread_id_seq | sequence | postgres
(17 rows)

openroom=# \dt
          List of relations
 Schema |  Name   | Type  |  Owner   
--------+---------+-------+----------
 public | board   | table | postgres
 public | media   | table | postgres
 public | post    | table | postgres
 public | poster  | table | postgres
 public | reply   | table | postgres
 public | session | table | postgres
 public | slip    | table | postgres
 public | tag     | table | postgres
 public | tags    | table | postgres
 public | thread  | table | postgres
(10 rows)

openroom=# \d+ public.*
                                                       Table "public.board"
 Column |         Type          | Collation | Nullable |              Default              | Storage  | Stats target | Description 
--------+-----------------------+-----------+----------+-----------------------------------+----------+--------------+-------------
 id     | integer               |           | not null | nextval('board_id_seq'::regclass) | plain    |              | 
 name   | character varying(32) |           | not null |                                   | extended |              | 
 rules  | character varying     |           |          |                                   | extended |              | 
Indexes:
    "board_pkey" PRIMARY KEY, btree (id)
Referenced by:
    TABLE "thread" CONSTRAINT "thread_board_fkey" FOREIGN KEY (board) REFERENCES board(id)

                    Sequence "public.board_id_seq"
  Type   | Start | Minimum |  Maximum   | Increment | Cycles? | Cache 
---------+-------+---------+------------+-----------+---------+-------
 integer |     1 |       1 | 2147483647 |         1 | no      |     1
Owned by: public.board.id

        Index "public.board_pkey"
 Column |  Type   | Definition | Storage 
--------+---------+------------+---------
 id     | integer | id         | plain
primary key, btree, for table "public.board"

                                                       Table "public.media"
 Column |         Type         | Collation | Nullable |              Default              | Storage  | Stats target | Description 
--------+----------------------+-----------+----------+-----------------------------------+----------+--------------+-------------
 id     | integer              |           | not null | nextval('media_id_seq'::regclass) | plain    |              | 
 ext    | character varying(3) |           | not null |                                   | extended |              | 
Indexes:
    "media_pkey" PRIMARY KEY, btree (id)
Referenced by:
    TABLE "post" CONSTRAINT "post_media_fkey" FOREIGN KEY (media) REFERENCES media(id) ON DELETE CASCADE

                    Sequence "public.media_id_seq"
  Type   | Start | Minimum |  Maximum   | Increment | Cycles? | Cache 
---------+-------+---------+------------+-----------+---------+-------
 integer |     1 |       1 | 2147483647 |         1 | no      |     1
Owned by: public.media.id

        Index "public.media_pkey"
 Column |  Type   | Definition | Storage 
--------+---------+------------+---------
 id     | integer | id         | plain
primary key, btree, for table "public.media"

                                                           Table "public.post"
  Column  |            Type             | Collation | Nullable |             Default              | Storage  | Stats target | Description 
----------+-----------------------------+-----------+----------+----------------------------------+----------+--------------+-------------
 id       | integer                     |           | not null | nextval('post_id_seq'::regclass) | plain    |              | 
 body     | character varying(4096)     |           | not null |                                  | extended |              | 
 subject  | character varying(64)       |           |          |                                  | extended |              | 
 thread   | integer                     |           | not null |                                  | plain    |              | 
 datetime | timestamp without time zone |           | not null |                                  | plain    |              | 
 poster   | integer                     |           | not null |                                  | plain    |              | 
 media    | integer                     |           |          |                                  | plain    |              | 
 spoiler  | boolean                     |           |          |                                  | plain    |              | 
Indexes:
    "post_pkey" PRIMARY KEY, btree (id)
Foreign-key constraints:
    "post_media_fkey" FOREIGN KEY (media) REFERENCES media(id) ON DELETE CASCADE
    "post_poster_fkey" FOREIGN KEY (poster) REFERENCES poster(id)
    "post_thread_fkey" FOREIGN KEY (thread) REFERENCES thread(id)
Referenced by:
    TABLE "reply" CONSTRAINT "reply_reply_from_fkey" FOREIGN KEY (reply_from) REFERENCES post(id)
    TABLE "reply" CONSTRAINT "reply_reply_to_fkey" FOREIGN KEY (reply_to) REFERENCES post(id)

                    Sequence "public.post_id_seq"
  Type   | Start | Minimum |  Maximum   | Increment | Cycles? | Cache 
---------+-------+---------+------------+-----------+---------+-------
 integer |     1 |       1 | 2147483647 |         1 | no      |     1
Owned by: public.post.id

        Index "public.post_pkey"
 Column |  Type   | Definition | Storage 
--------+---------+------------+---------
 id     | integer | id         | plain
primary key, btree, for table "public.post"

                                                         Table "public.poster"
   Column   |         Type         | Collation | Nullable |              Default               | Storage  | Stats target | Description 
------------+----------------------+-----------+----------+------------------------------------+----------+--------------+-------------
 id         | integer              |           | not null | nextval('poster_id_seq'::regclass) | plain    |              | 
 hex_string | character varying(4) |           | not null |                                    | extended |              | 
 ip_address | integer              |           | not null |                                    | plain    |              | 
 thread     | integer              |           | not null |                                    | plain    |              | 
 slip       | integer              |           |          |                                    | plain    |              | 
Indexes:
    "poster_pkey" PRIMARY KEY, btree (id)
Foreign-key constraints:
    "poster_slip_fkey" FOREIGN KEY (slip) REFERENCES slip(id)
    "poster_thread_fkey" FOREIGN KEY (thread) REFERENCES thread(id) ON DELETE CASCADE
Referenced by:
    TABLE "post" CONSTRAINT "post_poster_fkey" FOREIGN KEY (poster) REFERENCES poster(id)

                   Sequence "public.poster_id_seq"
  Type   | Start | Minimum |  Maximum   | Increment | Cycles? | Cache 
---------+-------+---------+------------+-----------+---------+-------
 integer |     1 |       1 | 2147483647 |         1 | no      |     1
Owned by: public.poster.id

       Index "public.poster_pkey"
 Column |  Type   | Definition | Storage 
--------+---------+------------+---------
 id     | integer | id         | plain
primary key, btree, for table "public.poster"

                                     Table "public.reply"
   Column   |  Type   | Collation | Nullable | Default | Storage | Stats target | Description 
------------+---------+-----------+----------+---------+---------+--------------+-------------
 reply_from | integer |           | not null |         | plain   |              | 
 reply_to   | integer |           | not null |         | plain   |              | 
Indexes:
    "reply_pkey" PRIMARY KEY, btree (reply_from, reply_to)
Foreign-key constraints:
    "reply_reply_from_fkey" FOREIGN KEY (reply_from) REFERENCES post(id)
    "reply_reply_to_fkey" FOREIGN KEY (reply_to) REFERENCES post(id)

          Index "public.reply_pkey"
   Column   |  Type   | Definition | Storage 
------------+---------+------------+---------
 reply_from | integer | reply_from | plain
 reply_to   | integer | reply_to   | plain
primary key, btree, for table "public.reply"

                                          Table "public.session"
 Column  |         Type          | Collation | Nullable | Default | Storage  | Stats target | Description 
---------+-----------------------+-----------+----------+---------+----------+--------------+-------------
 id      | character varying(32) |           | not null |         | extended |              | 
 slip_id | integer               |           | not null |         | plain    |              | 
Indexes:
    "session_pkey" PRIMARY KEY, btree (id)
Foreign-key constraints:
    "session_slip_id_fkey" FOREIGN KEY (slip_id) REFERENCES slip(id)

              Index "public.session_pkey"
 Column |         Type          | Definition | Storage  
--------+-----------------------+------------+----------
 id     | character varying(32) | id         | extended
primary key, btree, for table "public.session"

                                                         Table "public.slip"
  Column   |         Type          | Collation | Nullable |             Default              | Storage  | Stats target | Description 
-----------+-----------------------+-----------+----------+----------------------------------+----------+--------------+-------------
 id        | integer               |           | not null | nextval('slip_id_seq'::regclass) | plain    |              | 
 name      | character varying(20) |           | not null |                                  | extended |              | 
 pass_hash | character varying     |           | not null |                                  | extended |              | 
 is_admin  | boolean               |           | not null |                                  | plain    |              | 
 is_mod    | boolean               |           | not null |                                  | plain    |              | 
Indexes:
    "slip_pkey" PRIMARY KEY, btree (id)
    "slip_name_key" UNIQUE CONSTRAINT, btree (name)
Referenced by:
    TABLE "poster" CONSTRAINT "poster_slip_fkey" FOREIGN KEY (slip) REFERENCES slip(id)
    TABLE "session" CONSTRAINT "session_slip_id_fkey" FOREIGN KEY (slip_id) REFERENCES slip(id)

                    Sequence "public.slip_id_seq"
  Type   | Start | Minimum |  Maximum   | Increment | Cycles? | Cache 
---------+-------+---------+------------+-----------+---------+-------
 integer |     1 |       1 | 2147483647 |         1 | no      |     1
Owned by: public.slip.id

              Index "public.slip_name_key"
 Column |         Type          | Definition | Storage  
--------+-----------------------+------------+----------
 name   | character varying(20) | name       | extended
unique, btree, for table "public.slip"

        Index "public.slip_pkey"
 Column |  Type   | Definition | Storage 
--------+---------+------------+---------
 id     | integer | id         | plain
primary key, btree, for table "public.slip"

                                                       Table "public.tag"
   Column   |       Type        | Collation | Nullable |             Default             | Storage  | Stats target | Description 
------------+-------------------+-----------+----------+---------------------------------+----------+--------------+-------------
 id         | integer           |           | not null | nextval('tag_id_seq'::regclass) | plain    |              | 
 name       | character varying |           | not null |                                 | extended |              | 
 bg_style   | character varying |           |          |                                 | extended |              | 
 text_style | character varying |           |          |                                 | extended |              | 
Indexes:
    "tag_pkey" PRIMARY KEY, btree (id)
Referenced by:
    TABLE "tags" CONSTRAINT "tags_tag_id_fkey" FOREIGN KEY (tag_id) REFERENCES tag(id)

                     Sequence "public.tag_id_seq"
  Type   | Start | Minimum |  Maximum   | Increment | Cycles? | Cache 
---------+-------+---------+------------+-----------+---------+-------
 integer |     1 |       1 | 2147483647 |         1 | no      |     1
Owned by: public.tag.id

         Index "public.tag_pkey"
 Column |  Type   | Definition | Storage 
--------+---------+------------+---------
 id     | integer | id         | plain
primary key, btree, for table "public.tag"

                                     Table "public.tags"
  Column   |  Type   | Collation | Nullable | Default | Storage | Stats target | Description 
-----------+---------+-----------+----------+---------+---------+--------------+-------------
 tag_id    | integer |           | not null |         | plain   |              | 
 thread_id | integer |           | not null |         | plain   |              | 
Indexes:
    "tags_pkey" PRIMARY KEY, btree (tag_id, thread_id)
Foreign-key constraints:
    "tags_tag_id_fkey" FOREIGN KEY (tag_id) REFERENCES tag(id)
    "tags_thread_id_fkey" FOREIGN KEY (thread_id) REFERENCES thread(id)

          Index "public.tags_pkey"
  Column   |  Type   | Definition | Storage 
-----------+---------+------------+---------
 tag_id    | integer | tag_id     | plain
 thread_id | integer | thread_id  | plain
primary key, btree, for table "public.tags"

                                                             Table "public.thread"
    Column    |            Type             | Collation | Nullable |              Default               | Storage | Stats target | Description 
--------------+-----------------------------+-----------+----------+------------------------------------+---------+--------------+-------------
 id           | integer                     |           | not null | nextval('thread_id_seq'::regclass) | plain   |              | 
 board        | integer                     |           | not null |                                    | plain   |              | 
 views        | integer                     |           | not null |                                    | plain   |              | 
 last_updated | timestamp without time zone |           |          |                                    | plain   |              | 
Indexes:
    "thread_pkey" PRIMARY KEY, btree (id)
Foreign-key constraints:
    "thread_board_fkey" FOREIGN KEY (board) REFERENCES board(id)
Referenced by:
    TABLE "post" CONSTRAINT "post_thread_fkey" FOREIGN KEY (thread) REFERENCES thread(id)
    TABLE "poster" CONSTRAINT "poster_thread_fkey" FOREIGN KEY (thread) REFERENCES thread(id) ON DELETE CASCADE
    TABLE "tags" CONSTRAINT "tags_thread_id_fkey" FOREIGN KEY (thread_id) REFERENCES thread(id)

                   Sequence "public.thread_id_seq"
  Type   | Start | Minimum |  Maximum   | Increment | Cycles? | Cache 
---------+-------+---------+------------+-----------+---------+-------
 integer |     1 |       1 | 2147483647 |         1 | no      |     1
Owned by: public.thread.id

       Index "public.thread_pkey"
 Column |  Type   | Definition | Storage 
--------+---------+------------+---------
 id     | integer | id         | plain
primary key, btree, for table "public.thread"

```

**Comment by DangerOnTheRanger** [comment]
That schema looks about right to me - did you have any issues using Postgres? I do agree on the testing bit; I'll try to implement unit tests for new code I write going forwards starting with S3 storage.

Side note, I think at some point we should have config file support so we don't have to go into the source to modify stuff like the DB URL, but that can wait a while.

**Comment by 9034725985** [comment]
No, but I have no experience with slips. By the way, can you help me out on an unrelated project? I am trying to implement an existing (GPLv3) php application in flask 

https://github.com/openroom/openroom/tree/flask

**Comment by DangerOnTheRanger** [comment]
Slips are just username/password pairs the same as you might find with any regular website, except with Maniwani they're just intended to store user settings + identify admins and mods as such when the time arises.

Definitely feel free to tag me on any issues you need help with, but unfortunately I'm probably going to have to dial back the rapidity of my open-source contributions in general until I figure out what my workload for the semester starting next week looks like. Hopefully that won't take too long, though.

**Comment by DangerOnTheRanger** [comment]
Fixed as of bd0062cb03b010ce6db0eef8371c655ddeff7689.
